From faaf6a8246a64c623c664c22a55ccb626b7142e0 Mon Sep 17 00:00:00 2001
From: S-trace <S-trace@list.ru>
Date: Sun, 28 Feb 2016 09:03:54 +0300
Subject: [PATCH 2/2] [STORAGES] vold: Don't create volumes for service
 partitions

This fixes huge error notifications at every boot
on devices with broken eMMC

Change-Id: I68c9f00dd3cf470a1388b83c9a509540d0ab4b01
---
 Disk.cpp | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/Disk.cpp b/Disk.cpp
index 940cb70..502c688 100644
--- a/Disk.cpp
+++ b/Disk.cpp
@@ -78,6 +78,19 @@ static const char* kGptLinuxFilesystem = "0FC63DAF-8483-4772-8E79-3D69D8477DE4";
 static const char* kGptAndroidMeta = "19A710A2-B3CA-11E4-B026-10604B889DCF";
 static const char* kGptAndroidExpand = "193D1EA4-B3CA-11E4-B075-10604B889DCF";
 
+static const char* kServicePartitionNames[] = {
+    "sbl1bak",
+    "abootbak",
+    "rpmbak",
+    "tzbak",
+    "pad",
+    "modem",
+    "persist",
+    "cache",
+    "system",
+    "userdata"
+};
+
 enum class Table {
     kUnknown,
     kMbr,
@@ -361,6 +374,22 @@ status_t Disk::readPartitions() {
             } else if (table == Table::kGpt) {
                 const char* typeGuid = strtok(nullptr, kSgdiskToken);
                 const char* partGuid = strtok(nullptr, kSgdiskToken);
+                const char* partName = strtok(nullptr, kSgdiskToken);
+
+               unsigned int iter=0;
+               bool servicePartition=false;
+                for (iter=0; iter < (sizeof(kServicePartitionNames)/sizeof(kServicePartitionNames[0])); iter++) {
+                    if(!strcasecmp(partName, kServicePartitionNames[iter])) {
+                        servicePartition=true;
+                        break;
+                    }
+                else
+                    servicePartition=false;
+                }
+                if (servicePartition) {
+                    LOG(WARNING) << "Disk::readPartitions(): Skipping service partition " << partName;
+                    continue;
+                }
 
                 if (!strcasecmp(typeGuid, kGptBasicData)
                         || !strcasecmp(typeGuid, kGptLinuxFilesystem)) {
-- 
2.14.1

